copilot-theorem/src/Copilot/Theorem/What4/Translate.hs

```diff
@@ -687,42 +815,73 @@ translateOp2 origExpr sym op xe1 xe2 = case (op, xe1, xe2) of
+                         numCmp (WI.bvSgt sym) (WI.bvUgt sym)
                                 (\(_ :: WFP.FloatInfoRepr fi) -> WFP.iFloatGt @_ @fi sym)
                                 xe1 xe2
 
-  (CE.BwAnd _, xe1, xe2) -> bvOp (WI.bvAndBits sym) (WI.bvAndBits sym) xe1 xe2
-  (CE.BwOr  _, xe1, xe2) -> bvOp (WI.bvOrBits sym)  (WI.bvOrBits sym)  xe1 xe2
-  (CE.BwXor _, xe1, xe2) -> bvOp (WI.bvXorBits sym) (WI.bvXorBits sym) xe1 xe2
+  (CE.BwAnd _, xe1, xe2) -> liftIO $ bvOp (WI.bvAndBits sym) (WI.bvAndBits sym) xe1 xe2
+  (CE.BwOr  _, xe1, xe2) -> liftIO $ bvOp (WI.bvOrBits sym)  (WI.bvOrBits sym)  xe1 xe2
+  (CE.BwXor _, xe1, xe2) -> liftIO $ bvOp (WI.bvXorBits sym) (WI.bvXorBits sym) xe1 xe2
 
   -- Note: For both shift operators, we are interpreting the shifter as an
   -- unsigned bitvector regardless of whether it is a word or an int.
   -- TODO, does this match the intended semantics?
   (CE.BwShiftL _ _, xe1, xe2) -> do
-    Just (SomeBVExpr e1 w1 _ ctor1) <- return $ asBVExpr xe1
-    Just (SomeBVExpr e2 w2 _ _    ) <- return $ asBVExpr xe2
-    e2' <- case testNatCases w1 w2 of
-      NatCaseLT LeqProof -> WI.bvTrunc sym w1 e2
-      NatCaseEQ -> return e2
-      NatCaseGT LeqProof -> WI.bvZext sym w1 e2
-    ctor1 <$> WI.bvShl sym e1 e2'
+    Just (SomeBVExpr e1 w1 sgn1 ctor1) <- return $ asBVExpr xe1
+    Just (SomeBVExpr e2 w2 _    _    ) <- return $ asBVExpr xe2
+
+    e2' <- liftIO $ case testNatCases w1 w2 of
+        NatCaseLT LeqProof -> WI.bvTrunc sym w1 e2
+        NatCaseEQ -> return e2
+        NatCaseGT LeqProof -> WI.bvZext sym w1 e2
+    res <- liftIO $ WI.bvShl sym e1 e2'
+
+    -- The second argument should not be greater than or equal to the bit width
+    wBV <- liftIO $ WI.bvLit sym w1 $ BV.width w1
+    notTooLarge <- liftIO $ WI.bvUlt sym e2' wBV
+    addSidePred notTooLarge
+
+    case sgn1 of
+      Unsigned -> do
+        -- Non-zero bits should not be shifted out
+        otherDirection <- liftIO $ WI.bvLshr sym res e2'
+        noWrap <- liftIO $ WI.bvEq sym e1 otherDirection
+        addSidePred noWrap
+      Signed -> do
+        -- Bits that disagree with the sign bit should not be shifted out
+        otherDirection <- liftIO $ WI.bvAshr sym res e2'
+        noWrap <- liftIO $ WI.bvEq sym e1 otherDirection
+        addSidePred noWrap
+
+    return $ ctor1 res
   (CE.BwShiftR _ _, xe1, xe2) -> do
     Just (SomeBVExpr e1 w1 sgn1 ctor1) <- return $ asBVExpr xe1
     Just (SomeBVExpr e2 w2 _    _    ) <- return $ asBVExpr xe2
-    e2' <- case testNatCases w1 w2 of
+
+    e2' <- liftIO $ case testNatCases w1 w2 of
       NatCaseLT LeqProof -> WI.bvTrunc sym w1 e2
       NatCaseEQ -> return e2
       NatCaseGT LeqProof -> WI.bvZext sym w1 e2
-    ctor1 <$> case sgn1 of
+
+    -- The second argument should not be greater than or equal to the bit width
+    wBV <- liftIO $ WI.bvLit sym w1 $ BV.width w1
+    notTooLarge <- liftIO $ WI.bvUlt sym e2' wBV
+    addSidePred notTooLarge
```

Comment:
Besides this side condition, there's actually another possibility of poison arising if `lshr`/`ashr` is invoked with the `exact` flag. See [this corresponding code](https://github.com/GaloisInc/crucible/blob/9305587306de58e3cdd13009da8af856555deca3/crucible-llvm/src/Lang/Crucible/LLVM/Translation/Instruction.hs#L894-L897) on the `crucible-llvm` side. That being said, I cannot figure out how to convince Clang to emit `lshr`/`ashr` instructions with the `exact` flag at all, regardless of optimization level. I opted to leave this side condition out entirely.

Reply:
Indeed, I don't think non-exact shifting should be an error.